light-client/src/builder/light_client.rs

```diff
@@ -0,0 +1,213 @@
+//! DSL for building a light client [`Instance`]
+
+use tendermint::{block::Height, Hash};
+
+use crate::bail;
+use crate::builder::error::{self, Error};
+use crate::components::clock::Clock;
+use crate::components::io::{AtHeight, Io};
+use crate::components::scheduler::Scheduler;
+use crate::components::verifier::Verifier;
+use crate::light_client::{LightClient, Options};
+use crate::operations::Hasher;
+use crate::predicates::VerificationPredicates;
+use crate::state::{State, VerificationTrace};
+use crate::store::LightStore;
+use crate::supervisor::Instance;
+use crate::types::{LightBlock, PeerId, Status};
+
+#[cfg(feature = "rpc-client")]
+use {
+    crate::components::clock::SystemClock, crate::components::io::ProdIo,
+    crate::components::scheduler, crate::components::verifier::ProdVerifier,
+    crate::operations::ProdHasher, crate::predicates::ProdPredicates, std::time::Duration,
+    tendermint_rpc as rpc,
+};
+
+/// No trusted state has been set yet
+pub struct NoTrustedState;
+
+/// A trusted state has been set and validated
+pub struct HasTrustedState;
+
+/// Builder for a light client [`Instance`]
+#[must_use]
+pub struct LightClientBuilder<State> {
+    peer_id: PeerId,
+    options: Options,
+    io: Box<dyn Io>,
+    clock: Box<dyn Clock>,
+    hasher: Box<dyn Hasher>,
+    verifier: Box<dyn Verifier>,
+    scheduler: Box<dyn Scheduler>,
+    predicates: Box<dyn VerificationPredicates>,
+    light_store: Box<dyn LightStore>,
+
+    #[allow(dead_code)]
+    state: State,
+}
+
+impl<Current> LightClientBuilder<Current> {
+    /// Private method to move from one state to another
+    fn with_state<Next>(self, state: Next) -> LightClientBuilder<Next> {
+        LightClientBuilder {
+            peer_id: self.peer_id,
```

Comment:
Unfortunately not, using `..self` yields the following error because `self` does not have the right type:

```
error[E0308]: mismatched types
  --> light-client/src/builder/light_client.rs:53:39
   |
50 | impl<Current> LightClientBuilder<Current> {
   |      ------- found type parameter
51 |     /// Private method to move from one state to another
52 |     fn with_state<Next>(self, state: Next) -> LightClientBuilder<Next> {
   |                   ---- expected type parameter
53 |         LightClientBuilder { state, ..self }
   |                                       ^^^^ expected type parameter `Next`, found type parameter `Current`
   |
   = note: expected struct `builder::light_client::LightClientBuilder<Next>`
              found struct `builder::light_client::LightClientBuilder<Current>`
   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
```

Reply:
Lol ok. I think this and the need to have a dummy state field both motivate using the wrapper approach as I think that would solve both, but no biggie